types/src/v0/impls/l1.rs

```diff
@@ -172,6 +173,47 @@ impl L1Client {
         self.wait_for_block(number).await
     }
 
+    /// Get information about the first finalized block with timestamp greater than or equal
+    /// `timestamp`.
+    pub async fn wait_for_finalized_block_with_timestamp(&self, timestamp: U256) -> L1BlockInfo {
+        let interval = self.provider.get_interval();
+
+        // Sleep until approximately the right time for the desired block to appear.
+        let now = U256::from(OffsetDateTime::now_utc().unix_timestamp() as u64);
+        if timestamp > now {
+            let dur = (timestamp - now).as_u64();
+            tracing::warn!("sleeping for {dur:?}, until {timestamp}");
+            sleep(Duration::from_secs(dur)).await;
+        }
+
+        // Wait until the finalized block has timestamp greater or equal to `timestamp`.
+        let mut block = loop {
+            let Some(block) = self.get_finalized_block().await else {
+                sleep(interval).await;
+                continue;
+            };
+            if block.timestamp >= timestamp {
+                break block;
+            }
+            tracing::info!(
+                %timestamp,
+                ?block,
+                "waiting for finalized block with sufficient timestamp"
+            );
+            sleep(interval).await;
+        };
+
+        // The finalized block jumps by more than 1 at a time, so we might not have found the
+        // earliest block with the desired timestamp. Work backwards until we find it.
+        loop {
+            let prev = self.wait_for_block(block.number - 1).await;
+            if prev.timestamp < timestamp {
```

Comment:
if the prev.timestamp is > timestamp, shouldn't we break? as that means that the block we retrieved in the previous loop is the earliest block with the desired timestamp?

Reply:
We're iterating backwards. So if `prev.timestamp > timestamp`, it is possible that `prev.prev.timestamp > timestamp` as well, so we must keep going. This is why when we finally return, we return `block`, *not* `prev`: we keep going backwards until we see a block with timestamp which is *too small*, and then we return the last block we saw before that one

Reply:
I think `parent` is probably a less confusing name than `prev` here, it's the "parent" in the chain, not the block. from the "previous" iteration of the loop. I'll make that change

Reply:
ahh makes sense - got it!